### PR TITLE
Add texture array bindings

### PR DESCRIPTION
**Pull request by bch29**
Also adds support for texture arrays in shader reflection.

### CODE REVIEW

**Comment by bch29** [comment]
Since WGPU requires that we specify array size up front in the binding descriptor, but that information is not available in the shader, reflection requires some extra information from the programmer for the sizes of the arrays. That's what the new `ShaderReflectOptions` struct is for:

```
pub struct ShaderReflectOptions {
    pub bevy_conventions: bool,
    pub array_sizes: HashMap<String, NonZeroU32>,
}
```

`array_sizes` here is a map from binding name to number of array elements that will be bound.

File: pipelined/bevy_render2/src/shader/shader_reflect.rs [hunk 1]
```diff
@@ -395,20 +436,35 @@ mod tests {
...
+                                count: None,
                             },
-                            shader_stage: BindingShaderStage::VERTEX,
-                        }]
+                            BindingDescriptor {
+                                index: 1,
+                                name: "TextureArr".into(),
+                                bind_type: BindType::Texture {
+                                    multisampled: false,
+                                    view_dimension: TextureViewDimension::D2Array,
```
Comment (superdump) [inline pipelined/bevy_render2/src/shader/shader_reflect.rs]: I thought D2Array was for array textures not array of textures. That is, texture2DArray in a glsl shader.

Comment (bch29) [reply pipelined/bevy_render2/src/shader/shader_reflect.rs]: I think you are right. Weirdly I didn't get any vulkan validation errors with D2Array, but neither do I with D2. Updated.

**Comment by cart** [comment]
Closing this as we've removed the reflection code and the bevy_render2 RenderResourcContext abstractions

**Pull request closed without merging**